### PR TITLE
Add E0034 error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #24407.
cc @michaelsproul

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -169,24 +169,6 @@ match string {
...
-
-```
-match 5u32 {
-    // This range is ok, albeit pointless.
-    1 ... 1 => ...
-    // This range is empty, and the compiler can tell.
-    1000 ... 5 => ...
-}
-```
-"##,
```
Comment (nagisa) [inline src/librustc_typeck/diagnostics.rs]: Why is this removed?

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -211,6 +193,47 @@ Reference:
...
+impl Trait1 for Test { fn foo() {} }
+impl Trait2 for Test { fn foo() {} }
+
+fn main() {
+    Test::foo() // error, what foo() to call?
+}
+```
+
+To avoid this error, you have to keep only one of them and remove the others.
+So let's take our example and fix it:
```
Comment (nagisa) [inline src/librustc_typeck/diagnostics.rs]: This is more commonly and less intrusively fixed by using [UFCS](https://doc.rust-lang.org/book/ufcs.html).

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: That is quite a good question...

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -211,6 +211,47 @@ Reference:
 http://doc.rust-lang.org/reference.html#trait-objects
 "##,
 
+E0034: r##"
+The compiler doesn't know what method to call because more than one does
+have the same prototype. Bad example:
```
Comment (nagisa) [inline src/librustc_typeck/diagnostics.rs]: Ditto about “Bad”.

File: src/librustc_typeck/diagnostics.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Right, I didn't this. I'll add this example.

**Comment by GuillaumeGomez** [comment]
@nagisa: Does my UFCS example seems good to you?

**Comment by nagisa** [comment]
Yes, its OK.

**Comment by GuillaumeGomez** [comment]
Well, I guess we need to summon r+ now? ^^

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -211,6 +211,68 @@ Reference:
 http://doc.rust-lang.org/reference.html#trait-objects
 "##,
 
+E0034: r##"
+The compiler doesn't know what method to call because more than one does
+have the same prototype. Example:
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: s/one does have/one method has

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -211,6 +211,68 @@ Reference:
...
+
+trait Trait2 {
+    fn foo();
+}
+
+impl Trait1 for Test { fn foo() {} }
+impl Trait2 for Test { fn foo() {} }
+
+fn main() {
+    Test::foo() // error, what foo() to call?
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: s/what/which

File: src/librustc_typeck/diagnostics.rs [hunk 6]
```diff
@@ -211,6 +211,68 @@ Reference:
...
+}
+
+impl Trait1 for Test { fn foo() {} }
+
+fn main() {
+    Test::foo() // and now that's good!
+}
+```
+
+However, a better solution would be using Universal Function Call Syntax:
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: s/using Universal Function Call Syntax/using fully explicit naming of the type and trait

**Comment by nrc** [comment]
r=me with the small changes

**Comment by GuillaumeGomez** [comment]
@nrc: I updated, thanks for your help !
